### DISCUSSION TITLE
Waveshare USB to RS485 + mbusd

### DISCUSSION THREAD

**Discussion by jamesog** [nathanmarlor/foxess_modbus#235] [General]
I've just migrated from a direct LAN connection to using RS485. I grabbed a [Waveshare USB to RS485 (B)](https://www.waveshare.com/usb-to-rs485-b.htm) device, connected it to a Raspberry Pi and installed [mbusd](https://github.com/3cky/mbusd) to act as a TCP to RTU gateway. This is a much cheaper option for those have a spare Pi (or other device) lying around and are comfortable compiling and installing a service on a Linux system, compared to buying a ready-made TCP-RTU gateway.

The Waveshare device appears as `/dev/ttyACM0` in Linux.

`mbusd` is configured with:

<details>
<summary>/etc/mbusd/mbusd-ttyACM0.conf</summary>

```
$ cat /etc/mbusd/mbusd-ttyACM0.conf
########## Serial port settings #############

# Serial port device name
device = /dev/ttyACM0

# Serial port speed
speed = 9600

# Serial port mode
mode = 8n1

# RS-485 data direction control type (addc, rts, sysfs_0, sysfs_1)
trx_control = addc

# Sysfs file to use to control data direction
# trx_sysfile =

############# TCP port settings #############

# TCP server address to bind
address = 0.0.0.0

# TCP server port number
port = 502

# Maximum number of simultaneous TCP connections
maxconn = 32

# Connection timeout value in seconds
timeout = 60

######### Request/response settings #########

# Maximum number of request retries
retries = 3

# Pause between requests in milliseconds
pause = 100

# Response wait time in milliseconds
wait = 500

# Reply on Broadcast
replyonbroadcast = no
```

</details>

You enable it in systemd by running `sudo systemctl enable --now mbusd@ttyACM0`.

Everything is working great. I selected adapter type of "Other" in the integration and bumped it up to max read set at 50 with the default poll rate of 10s.

~~I don't seem to have the ability to set the charge periods though - there's no service available. My inverter is on a pretty old firmware (1.37) so I've opened a ticket with Fox asking them to upgrade it now that I'm no longer using the LAN connection. I didn't want to risk the upgrade before switching to RS485 so I didn't lose the LAN port. :-)~~
Edit: charge periods are available and do work, they're found under Developer Tools.

From reading #40 I tried to get some timings with modpoll, the results look promising.

<details>
<summary>50 registers read in around 166ms</summary>

```
$ time ./modpoll -m tcp -p 502 -a 247 -t 3 -0 -1 -c 50 -r 11000 172.16.66.12
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 50
Communication.........: 172.16.66.12, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
[11000]: 1889
[11001]: 26
[11002]: 493
[11003]: 31
[11004]: 0
[11005]: 0
[11006]: 1203
[11007]: -20
[11008]: -251
[11009]: 2477
[11010]: 10
[11011]: 198
[11012]: -30
[11013]: 269
[11014]: 5002
[11015]: 0
[11016]: 0
[11017]: 0
[11018]: 0
[11019]: 0
[11020]: 0
[11021]: 0
[11022]: 0
[11023]: 195
[11024]: 262
[11025]: 375
[11026]: 0
[11027]: 0
[11028]: 3000
[11029]: -3000
[11030]: 0
[11031]: 0
[11032]: 514
[11033]: 1285
[11034]: 1204
[11035]: 21
[11036]: 59
[11037]: 468
[11038]: 256
[11039]: 1296
[11040]: 1008
[11041]: 350
[11042]: 500
[11043]: 235
[11044]: 196
[11045]: 3346
[11046]: 3344
[11047]: 515
[11048]: 17
[11049]: 10782

real	0m0.166s
user	0m0.002s
sys	0m0.000s
```

</details>

<details>
<summary>The full 96 in around 264ms</summary>

```
$ time ./modpoll -m tcp -p 502 -a 247 -t 3 -0 -1 -c 96 -r 11000 172.16.66.12
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 96
Communication.........: 172.16.66.12, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
[11000]: 1858
[11001]: 24
[11002]: 454
[11003]: 31
[11004]: 0
[11005]: 0
[11006]: 1202
[11007]: -18
[11008]: -224
[11009]: 2452
[11010]: 10
[11011]: 185
[11012]: -31
[11013]: 257
[11014]: 5000
[11015]: 0
[11016]: 0
[11017]: 0
[11018]: 0
[11019]: 0
[11020]: 0
[11021]: -19
[11022]: 0
[11023]: 197
[11024]: 260
[11025]: 370
[11026]: 0
[11027]: 0
[11028]: 3000
[11029]: -3000
[11030]: 0
[11031]: 0
[11032]: 514
[11033]: 1285
[11034]: 1204
[11035]: 19
[11036]: 60
[11037]: 471
[11038]: 256
[11039]: 1296
[11040]: 1008
[11041]: 350
[11042]: 500
[11043]: 235
[11044]: 195
[11045]: 3345
[11046]: 3343
[11047]: 515
[11048]: 17
[11049]: 10794
[11050]: 60
[11051]: 0
[11052]: 0
[11053]: 0
[11054]: 86
[11055]: 0
[11056]: 2
[11057]: 1
[11058]: 1
[11059]: 1
[11060]: 1
[11061]: 0
[11062]: 0
[11063]: 0
[11064]: 0
[11065]: 0
[11066]: 0
[11067]: 0
[11068]: 0
[11069]: 0
[11070]: 1103
[11071]: 20
[11072]: 0
[11073]: 1461
[11074]: 34
[11075]: 0
[11076]: 1303
[11077]: 8
[11078]: 0
[11079]: 287
[11080]: 0
[11081]: 0
[11082]: 1644
[11083]: 42
[11084]: 0
[11085]: 1905
[11086]: 15
[11087]: 0
[11088]: 1278
[11089]: 28
[11090]: 0
[11091]: 1967
[11092]: 29
[11093]: 5893
[11094]: 3342
[11095]: 8485

real	0m0.264s
user	0m0.002s
sys	0m0.000s
```
</details>

One thing to note: after setting the Pi up, I deleted the old LAN integration from HA, re-added the integration, and once it started reading data again the energy usage readings went a bit wacky. I'm not sure if that will sort itself out after a while or not.

<img width="1394" alt="SCR-20230513-lffe" src="https://github.com/nathanmarlor/foxess_modbus/assets/982184/8aac6e55-ef4e-4997-83e3-7b195dd38e90">

**Comment by canton7** [comment]
Ooh very nice! I'll get that added as a supported device.

Odd that the charge period service isn't showing up. Do you have the time period sensors?

**Reply by canton7** [reply]
Wait, you mean the services don't show up at all, or they do but you can't select your inverter?

**Reply by jamesog** [reply]
Yep, I've got those sensors.

Maybe I'm missing something with the charge period service? I see this:

<img width="300" alt="image" src="https://github.com/nathanmarlor/foxess_modbus/assets/982184/13a38c18-d1d2-4f07-865b-f9de16dd6ec9">

Should there be both a device and service there? I'm kind of assuming it could be firmware related, hence I asked Fox to upgrade it.

**Reply by canton7** [reply]
Nah, services are a HA thing provided by the integration. They won't show up there - go to Developer Tools - Services and search for FoxESS

**Reply by canton7** [reply]
Weird the screenshot in the README shows 1 service, but it's always been "1 device" for me. Any ideas how you got it to show "1 service" @nathanmarlor?

**Reply by jamesog** [reply]
Ah! In that case yes, I have the service! And it looks like it works 🎉 

The screenshot in the README confused me, too. That's why I wondered if there was something not being registered properly. But, looks like it's all good. Thanks for the pointer!

**Reply by canton7** [reply]
We've got an open issue to write that up properly on the wiki at some point.

**Reply by nathanmarlor** [reply]
> Weird the screenshot in the README shows 1 service, but it's always been "1 device" for me. Any ideas how you got it to show "1 service" @nathanmarlor? 

@canton7 it's from early code I ripped out EM - it's set when the entity is registered iirc.

**Reply by nathanmarlor** [reply]
> Weird the screenshot in the README shows 1 service, but it's always been "1 device" for me. Any ideas how you got it to show "1 service" @nathanmarlor? 

@canton7 it's from early code I ripped out EM - it's set when the entities are registered iirc.
